File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/allergen/AllergenHelper.java
package openfoodfacts.github.scrachx.openfood.models.entities.allergen;

import org.apache.commons.collections.CollectionUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import openfoodfacts.github.scrachx.openfood.models.Product;

public class AllergenHelper {
    public static class Data {
        private final boolean incomplete;
        private final List<String> allergens;

        public Data(boolean incomplete, List<String> allergens) {
            this.incomplete = incomplete;
            this.allergens = allergens;
        }

        public boolean isIncomplete() {
            return incomplete;
        }

        public List<String> getAllergens() {
            return allergens;
        }

        public boolean isEmpty() {
            return !incomplete && CollectionUtils.isEmpty(allergens);
        }
    }

    private AllergenHelper() {
    }

    private static Data createEmpty() {
        return new Data(false, Collections.emptyList());
    }

    public static Data computeUserAllergen(Product product, List<AllergenName> userAllergens) {
        if (userAllergens.isEmpty()) {
            return createEmpty();
        }

        if (!product.getStatesTags().contains("en:ingredients-completed")) {
            return new Data(true, Collections.emptyList());
        }

        Set<String> productAllergens = new HashSet<>(product.getAllergensHierarchy());
        productAllergens.addAll(product.getTracesTags());

        Set<String> allergenMatch = new TreeSet<>();
        for (AllergenName allergenName : userAllergens) {
            if (productAllergens.contains(allergenName.getAllergenTag())) {
                allergenMatch.add(allergenName.getName());
            }
        }
        return new Data(false, Collections.unmodifiableList(new ArrayList<>(allergenMatch)));
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/allergen/AllergenHelper.kt
package openfoodfacts.github.scrachx.openfood.models.entities.allergen

import openfoodfacts.github.scrachx.openfood.models.Product
import openfoodfacts.github.scrachx.openfood.network.ApiFields
import org.jetbrains.annotations.Contract

object AllergenHelper {
    @Contract(" -> new")
    private fun createEmpty() = Data(false, emptyList())

    fun computeUserAllergen(product: Product, userAllergens: List<AllergenName>): Data {
        if (userAllergens.isEmpty()) return createEmpty()

        if (ApiFields.StateTags.INGREDIENTS_COMPLETED.tag !in product.statesTags)
            return Data(true, emptyList())

        val productAllergens = (product.allergensHierarchy + product.tracesTags).toSet()

        val matchingAllergens = userAllergens
            .filter { it.allergenTag in productAllergens }
            .map { it.name }
            .toSet()

        return Data(false, matchingAllergens.toList())
    }

    data class Data(val incomplete: Boolean, val allergens: List<String>) {
        fun isEmpty() = !incomplete && allergens.isEmpty()
    }
}